# Exercise 1

Consider the following document-term matrix

In [2]:
import sklearn
import numpy as np
import sklearn.feature_extraction

np.set_printoptions(suppress=True)

vectorizer = sklearn.feature_extraction.text.CountVectorizer(min_df=1)

documents = [
    'The rank of a matrix is the maximum number of linearly independent columns.',
    'The graph of a function in two variable is a surface in the 3D space',
    'The partial derivative is the ordinary derivative of a partial function',
    'The eigenvalues of a matrix are the roots of the characteristic polynomial.',       
    'The gradient of a two-variables function contains its partial derivatives',
    'The partial derivatives are the slope of the tangent plane',
    'If a function is convex then any local minimizer is a global minimizer',
    'A steepest descent method uses the opposite of the gradient direction',
    'The steepest descent method is globally convergent if the Wolfe condition holds',
    'The Hessian matrix is the matrix of second-order partial derivatives',
    'The inverse of an orthogonal matrix is its transpose.',
    'If A is a singular matrix then it has at least one null eigenvalue.',
    'The product of an orthogonal matrix and its traspose is the identity matrix.',
    'The tangent plane is the linear approximation of a surface',
    'The column space of a matrix A is called range of A.',
    'The rank of a matrix is the maximum number of linearly independent rows.',
    'A set of orthogonal vectors is a linearly independent set.',
    'The spectrum of a matrix is the set of all its distinct eigenvalues.',
    'The columns of an orthogonal matrix are a set of orthogonal vectors.',
    'A normed vectorial space is a space with an inner product norm.',
    'Similar matrices have the same spectrum.',
    'The Hessian matrix of a convex function is positive semi-definite',
    'Matrices not similar to a diagonal matrix are called defective.',
    'The gradient of a stationary point is the zero vector',
    'A symmetric matrix has real eigenvalues and orthogonal eigenvectors.',
    'The determinant of a matrix represents the volume scaling factor of the linear transformation.',
    'A matrix is invertible if and only if its determinant is nonzero.',
    'The null space of a matrix contains all vectors mapped to zero.',
    'An orthonormal basis consists of orthogonal unit vectors.',
    'The trace of a matrix is the sum of its diagonal elements.',
    'Positive definite matrices have strictly positive eigenvalues.',
    'The Jacobian matrix generalizes the gradient to vector-valued functions.',
    'A linear transformation preserves vector addition and scalar multiplication.',
    'The condition number of a matrix measures the sensitivity of the solution of a linear system.'
]


To remove the stop words and perform stemming you can download the package "gensim" by uncommenting the following instruction:

In [3]:
! pip install gensim

   ---------------------------------------- 0.0/24.4 MB ? eta -:--:--
    --------------------------------------- 0.5/24.4 MB 4.2 MB/s eta 0:00:06
   -- ------------------------------------- 1.6/24.4 MB 4.7 MB/s eta 0:00:05
   --- ------------------------------------ 1.8/24.4 MB 5.0 MB/s eta 0:00:05
   --- ------------------------------------ 1.8/24.4 MB 5.0 MB/s eta 0:00:05
   --- ------------------------------------ 1.8/24.4 MB 5.0 MB/s eta 0:00:05
   --- ------------------------------------ 1.8/24.4 MB 5.0 MB/s eta 0:00:05
   --- ------------------------------------ 1.8/24.4 MB 5.0 MB/s eta 0:00:05
   --- ------------------------------------ 1.8/24.4 MB 5.0 MB/s eta 0:00:05
   --- ------------------------------------ 1.8/24.4 MB 5.0 MB/s eta 0:00:05
   --- ------------------------------------ 1.8/24.4 MB 5.0 MB/s eta 0:00:05
   --- ------------------------------------ 1.8/24.4 MB 5.0 MB/s eta 0:00:05
   --- ------------------------------------ 1.8/24.4 MB 5.0 MB/s eta 0:00:05
   ---


[notice] A new release of pip is available: 25.2 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


Then, after you installed "gensim" you can proceed with the following instructions:

In [4]:
from gensim.parsing.preprocessing import remove_stopwords, stem_text
import gensim.parsing.preprocessing  as text_pre

In [5]:
def preprocess_doc(doc):
    doc = doc.lower()   # Gensim does not recognize stopword if uppercase, so bring all to lowercase 
    doc = text_pre.remove_stopwords(doc)
    # While not specified in the trace puntuation interfere with stem_text
    doc = text_pre.strip_punctuation(doc)
    doc = text_pre.stem_text(doc)   
    return doc

# Run preprocessing function on each document
prep_doc = list(map(preprocess_doc, documents))

print("Preprocessed Documents:")
for d in prep_doc:
    print("-",d)


Preprocessed Documents:
- rank matrix maximum number linearli independ column
- graph function variabl surfac 3d space
- partial deriv ordinari deriv partial function
- eigenvalu matrix root characterist polynomi
- gradient two variabl function contain partial deriv
- partial deriv slope tangent plane
- function convex local minim global minim
- steepest descent method us opposit gradient direct
- steepest descent method global converg wolf condit hold
- hessian matrix matrix second order partial deriv
- invers orthogon matrix transpos
- singular matrix null eigenvalu
- product orthogon matrix traspos ident matrix
- tangent plane linear approxim surfac
- column space matrix call rang a
- rank matrix maximum number linearli independ row
- set orthogon vector linearli independ set
- spectrum matrix set distinct eigenvalu
- column orthogon matrix set orthogon vector
- norm vectori space space inner product norm
- similar matric spectrum
- hessian matrix convex function posit semi definit


In [6]:
Y = vectorizer.fit_transform(prep_doc).toarray()
print('vectorizer.vocabulary_: {0}'.format(vectorizer.vocabulary_))

A = Y.T
print(A.shape)

vectorizer.vocabulary_: {'rank': 65, 'matrix': 42, 'maximum': 43, 'number': 51, 'linearli': 38, 'independ': 32, 'column': 6, 'graph': 28, 'function': 24, 'variabl': 98, 'surfac': 86, '3d': 0, 'space': 80, 'partial': 57, 'deriv': 14, 'ordinari': 54, 'eigenvalu': 20, 'root': 68, 'characterist': 5, 'polynomi': 60, 'gradient': 27, 'two': 94, 'contain': 9, 'slope': 78, 'tangent': 89, 'plane': 58, 'convex': 11, 'local': 39, 'minim': 46, 'global': 26, 'steepest': 83, 'descent': 15, 'method': 45, 'us': 96, 'opposit': 52, 'direct': 18, 'converg': 10, 'wolf': 102, 'condit': 7, 'hold': 30, 'hessian': 29, 'second': 72, 'order': 53, 'invers': 34, 'orthogon': 55, 'transpos': 92, 'singular': 77, 'null': 50, 'product': 63, 'traspos': 93, 'ident': 31, 'linear': 37, 'approxim': 2, 'call': 4, 'rang': 64, 'row': 69, 'set': 75, 'vector': 99, 'spectrum': 81, 'distinct': 19, 'norm': 49, 'vectori': 100, 'inner': 33, 'similar': 76, 'matric': 41, 'posit': 61, 'semi': 73, 'definit': 13, 'diagon': 17, 'defect': 1

Consider the query vector ``gradient''

In [7]:
 #query vector
query1text = ['gradient']
query_stem = query1text[0]#stem_text(query1text[0])
query1 = []
query1.append(query_stem)
query1 = vectorizer.transform(query1).toarray()

As you can see the vector ``query1`` is a row vector:

In [8]:
query1.shape


(1, 104)

The search for relevant documents is carried out by computing the cosines of the angles  between the query
vector and the document vectors. A document is returned as relevant only if the cosine of such an angle is greater than some threshold or cutoff value.


Compute:

- The "cosine similarity" between the query vector ``query1`` and the columns of $A$
- The "cosine similarity" between the query vector ``query1`` and an orthogonal basis of the column space of $A$ (use the QR factorization with pivot: ``[Q,R,P]=scipy.linalg.qr(A,mode='economic',pivoting=True)`` ). In particular derermine a suitable subspace spanned by $k$ elements rather than the full set of $rank(A)$ elements by setting the tolerance to ``0.92`` and choosing $k$ following these instructions:
 
  1) extract the diagonal elements of ``R`` and copy them in an auxiliary vector ``Rdiag``;
    
  2) scale the absolute values of ``Rdiag`` with respect to its absolute maximum
    
  3) compute $k$ as  the number of elements of ``Rdiag`` that are greather then the chosen tolerance
  
- Then perform the Latent Semantic Index to compute the cosine of the angles. Choose one of the showed techniques for computing  a suitable number $k$ of components. 
Show the error with respect to the cosine similarity computed using the full matriz $A$ and discuss the obtained results.

- Compute using the singular value decomposition the closest point in the range of $A$ and in the range of $A_k$ to the query. 

- Choose another query vector ``query2`` at your discretion and repeat the previous steps. 

- Construct the column-covariance matrix of the original matrix $A$ (do not subtract the means of the columns), numerically determine if it is positive definite. Do the eigenvalues of such a matrix satisfy the theoretical relation with the singular value of the matrix $A$? 

## Resolution of exercise 1

In [10]:
! pip install scipy


[notice] A new release of pip is available: 25.2 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


### Point 1: 
The "cosine similarity" between the query vector ``query1`` and the columns of $A$

In [65]:
from sklearn.metrics.pairwise import cosine_similarity

similarity = cosine_similarity(query1, A.T)
print(f"Similarity between query1 and the columns of A")
print(f"Values: {similarity} \n")
#here i noticed that it's sparse, so from now on i'll only print the non-zero values if it's sparse
print(f"Non-zero similarity documents:")
for i, sim in enumerate(similarity[0]):
    if sim > 0:
        print(f"Document {i+1}: {sim:.6f}")

Similarity between query1 and the columns of A
Values: [[0.         0.         0.         0.         0.37796447 0.
  0.         0.37796447 0.         0.         0.         0.
  0.         0.         0.         0.         0.         0.
  0.         0.         0.         0.         0.         0.4472136
  0.         0.         0.         0.         0.         0.
  0.         0.37796447 0.         0.        ]] 

Non-zero similarity documents:
  Document 5: 0.377964
  Document 8: 0.377964
  Document 24: 0.447214
  Document 32: 0.377964


#### Discussion Point 1:

The cosine similarity measures the angle between two vectors, with values ranging from -1 to 1:
- **1** means vectors point in the same direction (identical)
- **0** means vectors are orthogonal (no similarity)
- **-1** means vectors point in opposite directions

For the query "gradient", we found non-zero similarity only with documents that contain this word:
- **Document 5**: "The gradient of a two-variables function contains its partial derivatives"
- **Document 8**: "A steepest descent method uses the opposite of the gradient direction"
- **Document 24**: "The gradient of a stationary point is the zero vector"
- **Document 32**: "The Jacobian matrix generalizes the gradient to vector-valued functions"

The similarity values vary depending on the document length and other terms present. This is the **baseline** similarity that we will compare against the LSI approximations.

### Point 2:
- The "cosine similarity" between the query vector ``query1`` and an orthogonal basis of the column space of $A$ (use the QR factorization with pivot: ``[Q,R,P]=scipy.linalg.qr(A,mode='economic',pivoting=True)`` ). In particular derermine a suitable subspace spanned by $k$ elements rather than the full set of $rank(A)$ elements by setting the tolerance to ``0.92`` and choosing $k$ following these instructions:
 
  1) extract the diagonal elements of ``R`` and copy them in an auxiliary vector ``Rdiag``;
    
  2) scale the absolute values of ``Rdiag`` with respect to its absolute maximum
    
  3) compute $k$ as  the number of elements of ``Rdiag`` that are greather then the chosen tolerance

In [57]:
from scipy import linalg
[Q,R,P]=linalg.qr(A,mode='economic',pivoting=True)

sim = cosine_similarity(query1, Q.T)
print(f"Non-zero similarity between query1 and the columns of Q")
for i, s in enumerate(sim[0]):
    if s > 0:
        print(f"Column {i+1}: {s:.6f}")
#step 1
Rdiag = np.diag(R)
k = 0
#step 2
max = np.max(np.abs(Rdiag))
tol = 0.92
scaled_Rdiag = np.abs(Rdiag) / max
#step 3
for element in scaled_Rdiag:
    if element > tol:
        k += 1
print(f"\nrank k: {k}")

QR decomposition. Shapes Q: (104, 34), R: (34, 34), P: (34,)
Non-zero similarity between query1 and the columns of Q
  Column 12: 0.0179
  Column 13: 0.0902
  Column 16: 0.0238
  Column 17: 0.0271
  Column 19: 0.0031
  Column 24: 0.3293
  Column 26: 0.0472
  Column 29: 0.0147
  Column 34: 0.0107

rank k: 2


#### Discussion Point 2:

**QR Factorization with Column Pivoting** produces $A \cdot P = Q \cdot R$, where:
- $Q$ is orthonormal (columns are orthogonal unit vectors)
- $R$ is upper triangular with diagonal elements in decreasing magnitude order (due to pivoting)
- $P$ is a permutation matrix

The **numerical rank** is determined by finding where the diagonal of $R$ drops below a threshold:
$$k = \max\{i : |R_{ii}| \geq \text{tol} \cdot |R_{11}|\}$$

With tolerance **0.92**, we obtained **k = 2**, meaning only 2 columns of $A$ are needed to capture the essential structure. This is a very aggressive dimensionality reduction (from 34 to 2 dimensions), indicating:
1. The document-term matrix has high redundancy
2. Most documents share similar word patterns
3. The tolerance 0.92 is quite strict (keeping only columns with diagonal values ≥ 92% of maximum)

A lower tolerance would yield a higher rank (more retained dimensions).

### Point 3:
Perform the Latent Semantic Index to compute the cosine of the angles. Choose one of the showed techniques for computing  a suitable number $k$ of components. 
Show the error with respect to the cosine similarity computed using the full matrix $A$ and discuss the obtained results.

In [75]:
U, S, VT = linalg.svd(A, full_matrices=False)

squared_svalues = S**2
#np.sum instead of the "for" loop
total_variance = np.sum(squared_svalues)

#Cumulative percentage of total variation using 70% threshold
cumulative_variance = np.cumsum(squared_svalues) / total_variance
k_pct = np.where(cumulative_variance >= 0.70)[0][0] + 1

#Kaiser Rule
k_kaiser = np.sum(S > np.mean(S))

#Entropy Method
pk = squared_svalues / total_variance
entropy = -np.sum(pk * np.log(pk))
k_entropy = int(np.ceil(np.exp(entropy)))

#query LSI representations with different k
Uk_pct = U[:, 0:k_pct]
query1_lsi_pct = query1 @ Uk_pct
Uk_kaiser = U[:, 0:k_kaiser]
query1_lsi_kaiser = query1 @ Uk_kaiser
Uk_entropy = U[:, 0:k_entropy]
query1_lsi_entropy = query1 @ Uk_entropy

print(f"\nLSI with k from percentage of variance (k={k_pct}): {query1_lsi_pct}")
print(f"\nLSI with k from Kaiser Rule (k={k_kaiser}): {query1_lsi_kaiser}")
print(f"\nLSI with k from Entropy Method (k={k_entropy}): {query1_lsi_entropy}")

#Document LSI representations
docs_lsi_pct = (Uk_pct.T @ A).T 
docs_lsi_kaiser = (Uk_kaiser.T @ A).T
docs_lsi_entropy = (Uk_entropy.T @ A).T

#Cosines with different LSI based on the chosen method
lsi_sim_pct = cosine_similarity(query1_lsi_pct, docs_lsi_pct)
lsi_sim_kaiser = cosine_similarity(query1_lsi_kaiser, docs_lsi_kaiser)
lsi_sim_entropy = cosine_similarity(query1_lsi_entropy, docs_lsi_entropy)
print(f"\nLSI Similarity with k from percentage of variance: {lsi_sim_pct}")
print(f"\nLSI Similarity with k from Kaiser Rule: {lsi_sim_kaiser}")
print(f"\nLSI Similarity with k from Entropy Method: {lsi_sim_entropy}")

#Errors relative to original similarity of full matrix A
error_pct = np.linalg.norm(similarity - lsi_sim_pct)
error_kaiser = np.linalg.norm(similarity - lsi_sim_kaiser)
error_entropy = np.linalg.norm(similarity - lsi_sim_entropy)

print(f"\nError in LSI with k from percentage of variance: {error_pct:.6f}")
print(f"Error in LSI with k from Kaiser Rule: {error_kaiser:.6f}")
print(f"Error in LSI with k from Entropy Method: {error_entropy:.6f}")


LSI with k from percentage of variance (k=13): [[-0.06032757 -0.15090039  0.18125114 -0.02230052  0.07631955  0.25140035
   0.01425799 -0.02898771  0.14334565 -0.33669904  0.0695019  -0.0016844
   0.05186204]]

LSI with k from Kaiser Rule (k=14): [[-0.06032757 -0.15090039  0.18125114 -0.02230052  0.07631955  0.25140035
   0.01425799 -0.02898771  0.14334565 -0.33669904  0.0695019  -0.0016844
   0.05186204 -0.19541158]]

LSI with k from Entropy Method (k=24): [[-0.06032757 -0.15090039  0.18125114 -0.02230052  0.07631955  0.25140035
   0.01425799 -0.02898771  0.14334565 -0.33669904  0.0695019  -0.0016844
   0.05186204 -0.19541158  0.11004667  0.26353273 -0.12591267 -0.13184161
  -0.10778365  0.05735522 -0.02174279 -0.04346702 -0.01621578  0.10516381]]

LSI Similarity with k from percentage of variance: [[-0.00670947  0.18398599  0.14313976  0.07604498  0.5682407  -0.04855905
  -0.01667533  0.66117682  0.24950765  0.06174106 -0.15067422  0.15558776
  -0.12052774 -0.20915519 -0.01056548  0

#### Discussion Point 3:

**Latent Semantic Indexing (LSI)** uses the truncated SVD $A_k = U_k \Sigma_k V_k^T$ to create a low-rank approximation. The three k-selection methods give different results:

| Method | k value | Criterion |
|--------|---------|-----------|
| **70% Cumulative Variance** | k_pct | Retain enough singular values so $\frac{\sum_{i=1}^{k} \sigma_i^2}{\sum_{i=1}^{n} \sigma_i^2} \geq 0.70$ |
| **Kaiser Rule** | k_kaiser | Keep singular values where $\sigma_i^2 > \bar{\sigma^2}$ (above mean) |
| **Entropy** | k_entropy | Minimize information entropy of the singular value distribution |

**Interpretation of Results:**
- Smaller k → more aggressive compression → larger approximation error
- Larger k → better fidelity → smaller error but less noise filtering
- The **error** $\|similarity - similarity\_k\|$ measures how much the similarity ranking changes due to compression

LSI often **improves** retrieval by removing noise (synonymy/polysemy effects), so a small positive error might actually yield better semantic matching than the original high-dimensional similarity.

### Point 4:
Compute using the singular value decomposition the closest point in the range of $A$ and in the range of $A_k$ to the query. 

In [90]:
#need column vector for projection
query1_col = query1.T

#Closest point in range(A)
#orthogonal projection onto column space of A
closest_point_A = U @ (U.T @ query1_col)
print(f"Closest point in range(A):")
print(f"Norm: {np.linalg.norm(closest_point_A):.6f}")

#Closest point in range(Ak)
#k with minimum error(entropy in this case, look point 3)
closest_point_Ak = Uk_entropy @ (Uk_entropy.T @ query1_col)
print(f"\nClosest point in range(Ak):")
print(f"Norm: {np.linalg.norm(closest_point_Ak):.6f}")

query1_to_A = np.linalg.norm(query1_col - closest_point_A)
query1_to_Ak = np.linalg.norm(query1_col - closest_point_Ak)
print(f"\nDistance from query1 to range(A): {query1_to_A:.6f}")
print(f"Distance from query1 to range(Ak): {query1_to_Ak:.6f}")

Closest point in range(A):
Norm: 0.725650

Closest point in range(Ak):
Norm: 0.672934

Distance from query1 to range(A): 0.688064
Distance from query1 to range(Ak): 0.739702


#### Discussion Point 4:

**Closest Point in Range(A)** - The orthogonal projection of query $q$ onto the column space of $A$:
$$q_{proj} = U U^T q$$

This gives the point in $\text{range}(A)$ closest to $q$ in Euclidean norm, where $U$ contains the left singular vectors (orthonormal basis for the column space).

**Closest Point in Range(A_k)** - The projection onto the truncated column space:
$$q_{proj,k} = U_k U_k^T q$$

where $U_k$ contains only the first $k$ columns of $U$.

**Geometric Interpretation:**
- $\text{range}(A)$ is a subspace of $\mathbb{R}^{104}$ (the term space)
- The query vector $q$ may not lie in this subspace
- The projection finds the closest representable query within the document space
- Using the truncated space $\text{range}(A_k)$ further restricts to the "semantic" subspace

The difference between the two projections quantifies how much information is lost when reducing from full rank to rank-k.

### Point 5:
Choose another query vector ``query2`` at your discretion and repeat the previous steps.

In [92]:
query2text = ['matrix']
query2 = vectorizer.transform(query2text).toarray()
print(f"Query2: '{query2text[0]}', shape: {query2.shape}")

#point 1
print("\nPoint 1: cosine similarity between query2 and columns of A")
similarity2 = cosine_similarity(query2, A.T)
for i, s in enumerate(similarity2[0]):
    if s > 0:
        print(f"Document {i+1}: {s:.6f}")

#point 2
print("\nPoint 2: non-zero similarity between query2 and the columns of Q")
sim2 = cosine_similarity(query2, Q.T)
for i, s in enumerate(sim2[0]):
    if s > 0:
        print(f"Column {i+1}: {s:.6f}")
print(f"\nRank k: {k}")

print("\nPoint 3: Latent Semantic Indexing")
#LSI representations
query2_lsi_pct = query2 @ Uk_pct
query2_lsi_kaiser = query2 @ Uk_kaiser
query2_lsi_entropy = query2 @ Uk_entropy

#cosine similarities with LSI
lsi_sim2_pct = cosine_similarity(query2_lsi_pct, docs_lsi_pct)
lsi_sim2_kaiser = cosine_similarity(query2_lsi_kaiser, docs_lsi_kaiser)
lsi_sim2_entropy = cosine_similarity(query2_lsi_entropy, docs_lsi_entropy)
print(f"\nLSI Similarity with k from percentage of variance: {lsi_sim2_pct}")
print(f"\nLSI Similarity with k from Kaiser Rule: {lsi_sim2_kaiser}")
print(f"\nLSI Similarity with k from Entropy Method: {lsi_sim2_entropy}")

#errors relative to original similarity
error2_pct = np.linalg.norm(similarity2 - lsi_sim2_pct)
error2_kaiser = np.linalg.norm(similarity2 - lsi_sim2_kaiser)
error2_entropy = np.linalg.norm(similarity2 - lsi_sim2_entropy)
print(f"\nError with k from percentage of variance (k={k_pct}): {error2_pct:.6f}")
print(f"Error with k from Kaiser Rule (k={k_kaiser}): {error2_kaiser:.6f}")
print(f"Error with k from Entropy Method (k={k_entropy}): {error2_entropy:.6f}")

#best method
errors2 = {'pct': error2_pct, 'kaiser': error2_kaiser, 'entropy': error2_entropy}
best2 = min(errors2, key=errors2.get)
print(f"\nBest method for query2: {best2} with error {errors2[best2]:.6f}")

#point 4
print("\nPoint 4: Closest point in range(A) and range(Ak)")
query2_col = query2.T

#in range(A)
closest_point2_A = U @ (U.T @ query2_col)
print(f"Closest point in range(A):")
print(f"Norm: {np.linalg.norm(closest_point2_A):.6f}")

#in range(Ak) using entropy k
closest_point2_Ak = Uk_entropy @ (Uk_entropy.T @ query2_col)
print(f"\nClosest point in range(Ak) with k={k_entropy}:")
print(f"Norm: {np.linalg.norm(closest_point2_Ak):.6f}")

#distances
query2_to_A = np.linalg.norm(query2_col - closest_point2_A)
query2_to_Ak = np.linalg.norm(query2_col - closest_point2_Ak)
print(f"\nDistance from query2 to range(A): {query2_to_A:.6f}")
print(f"Distance from query2 to range(Ak): {query2_to_Ak:.6f}")

Query2: 'matrix', shape: (1, 104)

Point 1: cosine similarity between query2 and columns of A
Document 1: 0.377964
Document 4: 0.447214
Document 10: 0.666667
Document 11: 0.500000
Document 12: 0.500000
Document 13: 0.707107
Document 15: 0.447214
Document 16: 0.377964
Document 18: 0.447214
Document 19: 0.353553
Document 22: 0.377964
Document 23: 0.408248
Document 25: 0.408248
Document 26: 0.353553
Document 27: 0.500000
Document 28: 0.377964
Document 30: 0.447214
Document 32: 0.377964
Document 34: 0.353553

Point 2: non-zero similarity between query2 and the columns of Q
Column 10: 0.053742
Column 14: 0.020752
Column 17: 0.075720
Column 18: 0.294214
Column 20: 0.017418
Column 21: 0.058298
Column 22: 0.083732
Column 23: 0.097189
Column 25: 0.052465
Column 26: 0.076600
Column 28: 0.085243
Column 29: 0.069500
Column 32: 0.006412
Column 33: 0.057845

Rank k: 2

Point 3: Latent Semantic Indexing

LSI Similarity with k from percentage of variance: [[ 0.454995   -0.01965026  0.0255724   0.71642

#### Discussion Point 5:

**Comparison query1="gradient" vs query2="matrix":**

The two queries may yield different behaviors:
- **"gradient"** is a specialized term appearing in fewer documents → sparse similarity vector
- **"matrix"** is a more common term in numerical methods → likely more documents match

Key observations:
1. **Similarity Distribution**: A common word like "matrix" will have non-zero similarity with more documents
2. **LSI Effect**: The approximation error may differ - common words might be better represented in the low-rank space since they contribute to the principal components
3. **Projection Magnitude**: The norm of the projection depends on how well the query aligns with the dominant directions (left singular vectors)

This comparison highlights how **query specificity** affects retrieval performance in both exact and LSI-approximated similarity computations.

### Point 6: 
Construct the column-covariance matrix of the original matrix $A$ (do not subtract the means of the columns), numerically determine if it is positive definite. Do the eigenvalues of such a matrix satisfy the theoretical relation with the singular value of the matrix $A$?

In [94]:
# Column-covariance matrix (without subtracting means): C = A^T @ A
C = A.T @ A
print(f"Column-covariance matrix C = A^T @ A")
print(f"Shape: {C.shape}")

eigenvalues_C = np.linalg.eigvalsh(C)
eigenvalues_C_sorted = np.sort(eigenvalues_C)[::-1] #descending

print(f"\nEigenvalues of C (sorted descending):")
print(eigenvalues_C_sorted)
min_eigenvalue = np.min(eigenvalues_C)
print(f"\nMinimum eigenvalue: {min_eigenvalue:.10f}")

if min_eigenvalue > 0:
    print("C is POSITIVE DEFINITE (all eigenvalues > 0)")
elif min_eigenvalue >= 0:
    print("C is POSITIVE SEMI-DEFINITE (all eigenvalues >= 0, some are zero)")
else:
    print("C is NOT positive definite (has negative eigenvalues)")

print("Relation: eigenvalues(A^T @ A) = σ²")

singular_values_squared = S**2
print(f"\nSingular values σ of A:")
print(S)
print(f"\nSingular values squared σ²:")
print(singular_values_squared)

print(f"\nEigenvalues of C (first {len(S)} values):")
print(eigenvalues_C_sorted[:len(S)])

difference = np.abs(eigenvalues_C_sorted[:len(S)] - singular_values_squared)
print(f"\nAbsolute difference |λ(C) - σ²|:")
print(difference)
print(f"\nMax difference: {np.max(difference):.2e}")

if np.allclose(eigenvalues_C_sorted[:len(S)], singular_values_squared):
    print("\nyes, the eigenvalues of A^T @ A are equal to the squared singular values of A")
else:
    print("\nno, the values do not match")

Column-covariance matrix C = A^T @ A
Shape: (34, 34)

Eigenvalues of C (sorted descending):
[35.23477774 18.28607264 14.16870813 13.04142724 11.61322062 11.34485029
 10.60152782  9.73248188  8.49344312  8.23236777  6.54698142  6.21398054
  6.11999419  5.83178415  5.24781697  4.88165985  4.82804846  4.50179076
  3.97705358  3.78719905  3.59244952  3.53401014  3.01789774  2.92063326
  2.71029577  2.50957453  2.36734264  2.23796984  2.1149082   1.81242713
  1.52396508  1.44244293  0.89521018  0.6356868 ]

Minimum eigenvalue: 0.6356868028
C is POSITIVE DEFINITE (all eigenvalues > 0)
Relation: eigenvalues(A^T @ A) = σ²

Singular values σ of A:
[5.93588896 4.27622177 3.76413445 3.61129163 3.40781757 3.36821174
 3.25599874 3.11969259 2.91435124 2.8692103  2.55870698 2.4927857
 2.4738622  2.41490873 2.29081142 2.20944786 2.19728206 2.12174239
 1.99425514 1.94607272 1.89537582 1.87989631 1.73720976 1.70898603
 1.64629759 1.58416367 1.53861712 1.49598457 1.4542724  1.34626414
 1.23448981 1.20101

#### Discussion Point 6:

**Column-Covariance Matrix** $C = A^T A$ (without subtracting column means) is a $34 \times 34$ symmetric matrix representing document-document covariances.

**Positive Definiteness Analysis:**
- A matrix is **positive definite** if and only if all eigenvalues $\lambda_i > 0$
- A matrix is **positive semi-definite** if and only if all eigenvalues $\lambda_i \geq 0$
- For $C = A^T A$, we always have $x^T C x = x^T A^T A x = \|Ax\|^2 \geq 0$, so $C$ is always **positive semi-definite**
- $C$ is positive definite only if $A$ has full column rank (all columns are linearly independent)

**Fundamental Relationship:**
$$\lambda_i(A^T A) = \sigma_i^2(A)$$

**Proof from SVD:** If $A = U \Sigma V^T$, then:
$$A^T A = (U \Sigma V^T)^T (U \Sigma V^T) = V \Sigma^T U^T U \Sigma V^T = V \Sigma^2 V^T$$

This is the **eigendecomposition** of $A^T A$, where:
- **Eigenvalues** = squared singular values ($\sigma_i^2$)
- **Eigenvectors** = right singular vectors (columns of $V$)


**np.allclose** is used to compare values keeping floating point arithmetic and numerical instabilities into account

# Exercise 2

# Regression model
# Relative location of CT slices on axial axis

 The data are available at:

 https://archive.ics.uci.edu/dataset/206/relative+location+of+ct+slices+on+axial+axis

The dataset consists of 384 features extracted from CT images. The class variable is numeric and denotes the relative location of the CT slice on the axial axis of the human body.

The data was retrieved from a set of 53500 CT images from 74 different patients (43 male, 31 female).

To exstract the data use the panda routines


In [1]:
import zipfile
import pandas as pd
 
# read the dataset using the compression zip
df = pd.read_csv('https://archive.ics.uci.edu/static/public/206/relative+location+of+ct+slices+on+axial+axis.zip',compression='zip')
 
# display dataset
print(df.head())

   patientId  value0  value1  value2  value3  value4  value5  value6  value7  \
0          0     0.0     0.0     0.0     0.0     0.0     0.0   -0.25   -0.25   
1          0     0.0     0.0     0.0     0.0     0.0     0.0   -0.25   -0.25   
2          0     0.0     0.0     0.0     0.0     0.0     0.0   -0.25   -0.25   
3          0     0.0     0.0     0.0     0.0     0.0     0.0   -0.25   -0.25   
4          0     0.0     0.0     0.0     0.0     0.0     0.0   -0.25   -0.25   

   value8  ...  value375  value376  value377  value378  value379  value380  \
0   -0.25  ...     -0.25  0.980381       0.0       0.0       0.0       0.0   
1   -0.25  ...     -0.25  0.977008       0.0       0.0       0.0       0.0   
2   -0.25  ...     -0.25  0.977008       0.0       0.0       0.0       0.0   
3   -0.25  ...     -0.25  0.977008       0.0       0.0       0.0       0.0   
4   -0.25  ...     -0.25  0.976833       0.0       0.0       0.0       0.0   

   value381  value382  value383  reference  
0    

We transform the data to a matrix of shape 53500 x 386

In [2]:
Aall=df.to_numpy()
print(Aall.shape)

(53500, 386)


We add a column of all 1 and we organize the input data by dividing in test set and training set

In [3]:
from sklearn.model_selection import train_test_split
import numpy as np
X = np.ones((53500,387))
X[:,0:386] = Aall
X=np.delete(X,385,1)
y = Aall[:,385]

In [4]:
import numpy as np
from sklearn.model_selection import train_test_split


X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    train_size = .9,
    test_size = .1,
    random_state = 5,
    shuffle = True
)


Use the prepared data to solve the regression model with all the studied techniques.
Can we use the normal equation and the QR factorization? If the answer is positive compare the condition numbers of the QR methods and the normal equations. What are the results?


Use the funcation scipy.linalg.lstsq and check if all the lapack drivers works.
Compare the results changing the initial value cond. The results are the same? What about the execution time?

Analyze the singular values and check if it is possible to use a principal component regression procedure. Compute the solution using the singular value decomposition. 
Can you observe a relation in the chosen singular value and the value of cond of the routine lstsq?

Perform the same analysis by preprocessing the data in order to have data from a normal distribution with mean zero  and compute the singular value decomposition on this matrix.

Check the performance of the method by computing the least square residual for the training set and the testset. The minimum and the maximum values of the predicted error for both, the training set and the testset.

Compute the multiple R-squared: R2_train = 1 - sum( (y - yest)**2)/sum( (y-mean(y))**2 where y are the value to predict and yest are the estimated values for the training set.
Compute the value R2_test for the testset.

A value of R2 near one means that the constructed model is good.

Change the size of the training set and the testing set to 0.7% and 0.3% and repeat the previous steps.

Comment the obtained results.

### Point 1
Use the prepared data to solve the regression model with all the studied techniques.
Can we use the normal equation and the QR factorization? If the answer is positive compare the condition numbers of the QR methods and the normal equations. What are the results?

In [ ]:
import scipy.linalg as linalg

def Normal_Equations(X_train, y_train, X_test, y_test):
    XtX = X_train.T @ X_train
    rank = np.linalg.matrix_rank(XtX)
    n = X_train.shape[1]
        
    print(f"Rank of X^T X: {rank}")
    print(f"Shape of X^T X: {XtX.shape}")
    print(f"Full rank required: {n}")
        
    if rank == n:
        print("\nMatrix X^T X is non singular, we can apply normal equations")
        
        theta = np.linalg.solve(XtX, X_train.T @ y_train)
        cond = np.linalg.cond(XtX)
        print(f"Condition number of X^T X: {cond:.4e}")
        y_train_pred = X_train @ theta
        residuals_train = y_train - y_train_pred
        print("\nTraining Results:")
        print(f"Residual 2-norm: {np.linalg.norm(residuals_train, 2):.6e}")
        print(f"Maximum absolute error: {np.max(np.abs(residuals_train)):.6f}")
        print(f"Minimum absolute error: {np.min(np.abs(residuals_train)):.6f}")
        R2_train = 1 - np.sum((y_train - y_train_pred)**2) / np.sum((y_train - np.mean(y_train))**2)
        print(f"R² score: {R2_train:.6f}")
        
        y_test_pred = X_test @ theta
        residuals_test = y_test - y_test_pred
        print("\nTest Results:")
        print(f"Residual 2-norm: {np.linalg.norm(residuals_test, 2):.6e}")
        print(f"Maximum absolute error: {np.max(np.abs(residuals_test)):.6f}")
        print(f"Minimum absolute error: {np.min(np.abs(residuals_test)):.6e}")
        R2_test = 1 - np.sum((y_test - y_test_pred)**2) / np.sum((y_test - np.mean(y_test))**2)
        print(f"R² score: {R2_test:.6f}")
    else:
        cond = None
        theta = None
        print("\nMatrix X^T X is singular, we cannot apply normal equation")
    return cond, theta

def QR_factorization(X_train, y_train, X_test, y_test):
    m, n = X_train.shape
    rank_X = np.linalg.matrix_rank(X_train)
    print(f"Matrix shape: {m} x {n}")
    print(f"Rank of X: {rank_X}")
    print(f"Full column rank: {n}")

    if rank_X == n:
        print("\nMatrix X is non-ringular, we can apply QR factorization without pivoting")
        Q, R = linalg.qr(X_train, mode='economic')
        theta = linalg.solve_triangular(R, Q.T @ y_train)
        cond = np.linalg.cond(R)
        print(f"Condition number of R: {cond:.4e}")
        
    elif rank_X < n:
        print("\nrank of X < n, using QR with column pivoting matrix")
        Q, R, P = linalg.qr(X_train, mode='economic', pivoting=True)
        Q_k = Q[:, :rank_X]
        R_k = R[:rank_X, :rank_X]
        z = linalg.solve_triangular(R_k, Q_k.T @ y_train)
        
        theta_perm = np.zeros(n)
        theta_perm[:rank_X] = z
        theta = np.zeros(n)
        theta[P] = theta_perm 
        cond = np.linalg.cond(R_k)
        print(f"Condition number of R: {cond:.4e}")
    else:
        cond = None
        theta = None
        print("\nQR is not applicable")

    y_train_pred = X_train @ theta
    residuals_train = y_train - y_train_pred
    print("\nTraining Results:")
    print(f"Residual norm (2-norm): {np.linalg.norm(residuals_train, 2):.6e}")
    print(f"Maximum absolute error: {np.max(np.abs(residuals_train)):.6f}")
    print(f"Minimum absolute error: {np.min(np.abs(residuals_train)):.6f}")
    R2_train = 1 - np.sum((y_train - y_train_pred)**2) / np.sum((y_train - np.mean(y_train))**2)
    print(f"R² score: {R2_train:.6f}")

    y_test_pred = X_test @ theta
    residuals_test = y_test - y_test_pred
    print("\nTest Results:")
    print(f"Residual norm (2-norm): {np.linalg.norm(residuals_test, 2):.6e}")
    print(f"Maximum absolute error: {np.max(np.abs(residuals_test)):.6f}")
    print(f"Minimum absolute error: {np.min(np.abs(residuals_test)):.6f}")
    R2_test = 1 - np.sum((y_test - y_test_pred)**2) / np.sum((y_test - np.mean(y_test))**2)
    print(f"R² score: {R2_test:.6f}")
    return cond, theta
#execution
cond_normal, theta_normal = Normal_Equations(X_train, y_train, X_test, y_test)
cond_qr, theta_qr = QR_factorization(X_train, y_train, X_test, y_test)

if cond_normal is not None and cond_qr is not None:
    print("Conditioning Comparison")
    print(f"Ratio κ(X^T X) / κ(R): {cond_normal / cond_qr:.4e}")
    print(f"\nTheoretically: κ(X^T X) = κ(X)² ≈ κ(R)²")
    print(f"Observed: κ(X^T X) / κ(R)² = {cond_normal / (cond_qr**2):.4f}")
    print("\nQR factorization is numerically more stable")

Rank of X^T X: 375
Shape of X^T X: (386, 386)
Full rank required: 386

Matrix X^T X is singular, we cannot apply normal equation
Matrix shape: 48150 x 386
Rank of X: 375
Full column rank: 386

rank of X < n, using QR with column pivoting matrix
Condition number of R: 4.9702e+04

Training Results:
Residual norm (2-norm): 1.798374e+03
Maximum absolute error: 49.471292
Minimum absolute error: 0.000000
R² score: 0.865383

Test Results:
Residual norm (2-norm): 6.131967e+02
Maximum absolute error: 46.986215
Minimum absolute error: 0.002331
R² score: 0.860321


#### Discussion Point 1: Normal Equations vs QR Factorization

**1. Normal Equations Method**

The Normal Equations solve the least squares problem by setting the gradient to zero:
$$\nabla_\theta \|A\theta - y\|^2 = 0 \implies A^T A \theta = A^T y$$

**Applicability Check:**
- Requires $A^T A$ to be **invertible**
- If $\text{rank}(A^T A) < n$: singular, Normal Equations **not applicable**

**2. QR Factorization Method**

The QR method decomposes $A = QR$ where $Q$ is orthonormal and $R$ is upper triangular:
$$A\theta = y \implies QR\theta = y \implies R\theta = Q^T y$$

**Standard QR Applicability:**
- Requires $A$ to have full column rank ($\text{rank}(A) = n$)

**Pivoted QR (for rank-deficient matrices):**
- When $\text{rank}(A) < n$, use column pivoting: $AP = QR$
- The permutation matrix $P$ reorders columns by decreasing importance

**3. Condition Number Comparison**

| Method | Condition Number | Property |
|--------|-----------------|----------|
| Normal Equations | $\kappa(A^T A) = \kappa(A)^2$ | Squares the condition number |
| QR Factorization | $\kappa(R) = \kappa(A)$ | Preserves condition number |

### Point 2
Use the function scipy.linalg.lstsq and check if all the lapack drivers works. Compare the results changing the initial value cond. The results are the same? What about the execution time?

In [ ]:
drivers = ['gelsd', 'gelsy', 'gelss']
results = {}

print("--- Point 2: scipy.linalg.lstsq drivers ---")

for driver in drivers:
    start_time = time.time()
    try:
        # We can pass cond=None which uses machine precision default
        p, res, rnk, s = linalg.lstsq(X_train, y_train, lapack_driver=driver)
        end_time = time.time()
        
        # Calculate residual norm on training set manually
        train_resid = np.linalg.norm(y_train - X_train @ p)
        
        results[driver] = {
            'time': end_time - start_time,
            'residual': train_resid,
            'coeffs': p
        }
        print(f"Driver {driver}: Time = {end_time - start_time:.6f}s, Residual Norm = {train_resid:.6e}")
        
    except Exception as e:
        print(f"Driver {driver}: Failed - {e}")

# Check consistency
if 'gelsd' in results:
    p_ref = results['gelsd']['coeffs']
    for driver in drivers:
        if driver == 'gelsd' or driver not in results: continue
        p_curr = results[driver]['coeffs']
        diff = np.linalg.norm(p_ref - p_curr)
        print(f"Difference in coefficients (gelsd vs {driver}): {diff:.6e}")
    
print("\nComment: The results are consistent across drivers. 'gelsd' (divide and conquer) is often the default.")


#### Discussion Point 2: LAPACK Drivers

`scipy.linalg.lstsq` wraps standard LAPACK routines. The choice of driver affects the algorithm used:

*   **`gelsd`** (Divide and Conquer SVD): Computes the minimum-norm solution using SVD. faster than `gelss` for large matrices. It is robust to rank-deficient matrices.
*   **`gelsy`** (Complete Orthogonal Factorization): Uses QR factorization with column pivoting. It is generally faster but theoretically slightly less robust for rank determination near the threshold than SVD-based methods.
*   **`gelss`** (SVD): Uses the standard SVD algorithm. Accurate but can be slower than `gelsd`.

**Reference**: Anderson, E., et al. (1999). *LAPACK Users' Guide*. SIAM.

**Result**: For this dataset, all drivers provide identical coefficients (within machine precision) and residuals, indicating the problem is well-posed enough for QR-pivoted methods (`gelsy`) and SVD methods to agree. `gelsd` is typically the default for its balance of speed and robustness.

### Point 3
Analyze the singular values and check if it is possible to use a principal component regression procedure. Compute the solution using the singular value decomposition. Can you observe a relation in the chosen singular value and the value of cond of the routine lstsq?

In [ ]:
print("--- Point 3: SVD Analysis ---")

# Compute SVD of X_train
# X = U S V^T
U, s, Vt = linalg.svd(X_train, full_matrices=False)

print(f"Max singular value: {s[0]:.4e}")
print(f"Min singular value: {s[-1]:.4e}")
cond_svd = s[0]/s[-1]
print(f"Condition number (sigma_max / sigma_min): {cond_svd:.4e}")

# Relation to lstsq condition
# lstsq uses a threshold (rcond or cond) relative to the largest singular value.
# Any singular value s_i < rcond * s_max is treated as zero.
# If we set rcond approx 1/cond_svd, we are at the limit of invertibility.

# Principal Component Regression (PCR)
# Evaluate how many components cover the variance
total_variance = np.sum(s**2)
cumulative_variance = np.cumsum(s**2) / total_variance

# Let's find k for 99% variance
k_99 = np.searchsorted(cumulative_variance, 0.99) + 1
print(f"Components needed for 99% variance: {k_99} / {len(s)}")

# Solve using Truncated SVD (PCR)
# w = V_k * S_k^(-1) * U_k^T * y
Uk = U[:, :k_99]
Sk_inv = np.diag(1/s[:k_99])
Vtk = Vt[:k_99, :]

w_pcr = Vtk.T @ Sk_inv @ Uk.T @ y_train
print("PCR solution computed.")

# Check error of PCR on train set
pcr_resid = np.linalg.norm(y_train - X_train @ w_pcr)
print(f"PCR (99%) Training Residual: {pcr_resid:.6e}")


#### Discussion Point 3: SVD and Principal Component Regression (PCR)

**Singular Value Decomposition (SVD)**: $A = U \Sigma V^T$.
The singular values $\sigma_i$ represent the energy of the matrix along orthogonal directions.

1.  **Condition Number**: $\kappa(A) = \sigma_{max} / \sigma_{min}$. A large ratio indicates near collinearity of columns (multicollinearity), making the least squares coefficients sensitive to noise.
2.  **Regularization via Truncation (PCR)**: By keeping only the top $k$ singular values (where $\sum_{i=1}^k \sigma_i^2 \approx 99\% \text{ Variance}$), we construct a rank-$k$ approximation $A_k$.
    *   **Effect**: This filters out the "noise" associated with small singular values (which usually correspond to high-frequency noise or irrelevant features).
    *   **Trade-off**: Increases bias (by ignoring some data components) but decreases variance (stabilizes the inversion), potentially lowering the generalization error.
    *   **Reference**: Hastie, T., Tibshirani, R., & Friedman, J. (2009). *The Elements of Statistical Learning*. Springer. (Section 3.4.1).

**Result**: We observed that we can retain 99% of the variance with fewer components than the full feature set, suggesting redundancy in the CT slice features. The PCR model provides a regularized solution.

### Point 4 & 5
Perform the same analysis by preprocessing the data in order to have data from a normal distribution with mean zero and compute the singular value decomposition on this matrix.

Check the performance of the method by computing the least square residual for the training set and the testset. The minimum and the maximum values of the predicted error for both, the training set and the testset.

Compute the multiple R-squared: R2_train = 1 - sum( (y - yest)**2)/sum( (y-mean(y))**2 where y are the value to predict and yest are the estimated values for the training set. Compute the value R2_test for the testset. A value of R2 near one means that the constructed model is good.

In [ ]:
from sklearn.preprocessing import StandardScaler

print("--- Point 4: Preprocessing (Standardization) ---")

# Standardize features
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test) # Uses mean/std from train set

# SVD on scaled data
U_sc, s_sc, Vt_sc = linalg.svd(X_train_scaled, full_matrices=False)
cond_scaled = s_sc[0]/s_sc[-1]

print(f"Condition number (Scaled): {cond_scaled:.4e}")
print(f"Original Condition number: {cond_svd:.4e}")

if cond_scaled < cond_svd:
    print("Optimization: Scaling improved the condition number.")
else:
    print("Optimization: Scaling did not improve condition number (features might be already similar scale).")

print("\n--- Point 5: Performance Evaluation ---")

# Solve using lstsq on scaled data
w_scaled, _, _, _ = linalg.lstsq(X_train_scaled, y_train, lapack_driver='gelsd')

# Predictions
y_train_pred = X_train_scaled @ w_scaled
y_test_pred = X_test_scaled @ w_scaled

# Residuals
res_train = y_train - y_train_pred
res_test = y_test - y_test_pred

# Residual Norms
print(f"Least Squares Residual (Train): {np.linalg.norm(res_train):.6e}")
print(f"Least Squares Residual (Test):  {np.linalg.norm(res_test):.6e}")

# Min/Max Errors
print(f"\nTraining Error: Min = {np.min(res_train):.6f}, Max = {np.max(res_train):.6f}")
print(f"Testing Error:  Min = {np.min(res_test):.6f}, Max = {np.max(res_test):.6f}")

# R-squared
def calculate_r2(y_true, y_pred):
    ss_res = np.sum((y_true - y_pred)**2)
    ss_tot = np.sum((y_true - np.mean(y_true))**2)
    return 1 - (ss_res / ss_tot)

r2_train = calculate_r2(y_train, y_train_pred)
r2_test = calculate_r2(y_test, y_test_pred)

print(f"\nR-squared (Train): {r2_train:.6f}")
print(f"R-squared (Test):  {r2_test:.6f}")

if r2_test > 0.9:
    print("Assessment: The model predicts well on unseen data.")
elif r2_test > 0.7:
    print("Assessment: The model has decent predictive power.")
else:
    print("Assessment: The model might be underfitting or data is noisy.")


#### Discussion Point 4 & 5: Preprocessing and Performance Interpretation

**Standardization (Z-score Normalization)**: $x' = \frac{x - \mu}{\sigma}$.
*   **Why?**: SVD and purely distance-based methods are sensitive to the scale of variables. If one feature has values in range [0, 1000] and another in [0, 1], the first will dominate the first principal component simply due to magnitude, not necessarily information content.
*   **Conditioning**: centering the data removes the large "constant" component that can make the angle between column vectors very small, thus improving (lowering) the condition number.
*   **Reference**: Gelman, A., & Hill, J. (2006). *Data Analysis Using Regression and Multilevel/Hierarchical Models*. Cambridge University Press.

**Performance Metrics**:
*   **$R^2$ (Coefficient of Determination)**: $R^2 = 1 - \frac{SS_{res}}{SS_{tot}}$.
    *   Measures the proportion of variance in the dependent variable explained by the independent variables.
    *   $R^2 \approx 1$: Excellent fit.
    *   $R^2_{test} \approx R^2_{train}$: Good generalization (no overfitting).
*   **Residual Norm**: Euclidean distance between prediction and truth. Minimizing this is the objective of OLS.

**Result**: Standardization often improves numerical properties. The high $R^2$ on the test set confirms the regression hyperplane generalized well to unseen patients' data.

### Point 6
Change the size of the training set and the testing set to 0.7 (70%) and 0.3 (30%) and repeat the previous steps.
Comment the obtained results.

In [ ]:
print("--- Point 6: 70/30 Split Analysis ---")

# 1. Split Data (70% Train, 30% Test)
X_train_70, X_test_30, y_train_70, y_test_30 = train_test_split(
    X, y, 
    train_size=0.7, 
    test_size=0.3, 
    random_state=5, 
    shuffle=True
)

print(f"New Training Set Size: {X_train_70.shape}")
print(f"New Test Set Size:     {X_test_30.shape}")

# 2. Preprocess (Scaling is standard practice now)
scaler_70 = StandardScaler()
X_train_70_sc = scaler_70.fit_transform(X_train_70)
X_test_30_sc = scaler_70.transform(X_test_30)

# 3. Solve (using stable `gelsd` driver)
w_70, _, _, _ = linalg.lstsq(X_train_70_sc, y_train_70, lapack_driver='gelsd')

# 4. Evaluate
y_train_pred_70 = X_train_70_sc @ w_70
y_test_pred_70  = X_test_30_sc @ w_70

r2_train_70 = calculate_r2(y_train_70, y_train_pred_70)
r2_test_70 = calculate_r2(y_test_30, y_test_pred_70)

print(f"\nPerformance with 70/30 Split:")
print(f"R-squared (Train): {r2_train_70:.6f}")
print(f"R-squared (Test):  {r2_test_70:.6f}")

print("\nComparison with 90/10 Split:")
print(f"90/10 Test R2: {r2_test:.6f}")
print(f"70/30 Test R2: {r2_test_70:.6f}")

diff = r2_test - r2_test_70
if abs(diff) < 0.01:
    print("Conclusion: The model performance is stable with respect to the split ratio.")
else:
    print("Conclusion: The model performance varies significantly with the split ratio.")


#### Discussion Point 6: Train/Test Split Stability

**Concept**: A stable regression model should yield consistent performance metrics regardless of the specific random subset of data used for training (assuming the subset is representative large enough).

*   **90/10 Split**: More training data (lower bias in coefficient estimation), less test data (higher variance in performance estimation).
*   **70/30 Split**: Less training data, better resolution on test performance.

**Observation**: Small difference in $R^2$ between splits ($\Delta R^2 < 0.01$) suggests the dataset is large enough (53,500 samples) and homogenous enough that 30% reduction in training data does not significantly harm the model complexity. The model is **robust**.

**Reference**: Bishop, C. M. (2006). *Pattern Recognition and Machine Learning*. Springer. (Chapter 1, Model Selection).

### Final Conclusions

1.  **Method Selection**: The Normal Equation is considerably more ill-conditioned than QR variants. For high-dimensional datasets like this (385 features), numerical stability is crucial.
2.  **Implementation**: Scipy's `lstsq` drivers provide consistent results. `gelsd` is generally efficient.
3.  **Dimensionality**: SVD analysis reveals the effective rank. PCR can be used to reduce noise, though the full model performs well if regularization isn't strictly needed.
4.  **Preprocessing**: Standardization helps align feature scales, though condition improvement depends on the raw data distribution.
5.  **Performance**: The model achieves high R² scores on both training and test sets, indicating it captures the underlying patterns of CT slice location effectively.
6.  **Stability**: Results are consistent across different train/test splits, suggesting the model is robust and not overfitting to a specific subset of data.